### DISCUSSION TITLE
v1 and v2 snippets and templates

### DISCUSSION THREAD

**Discussion by fade2gray** [mark-wiemer/ahkpp#409] [Ideas]
Finally figured how to create a custom v2 template as prescribed [here](https://github.com/mark-wiemer-org/ahkpp/blob/main/docs/TemplateSnippet.md), and adjusted the template name in the settings.
```JSON
{
    "AhkV2Template": {
        "prefix": "#TEMPLATEv2",
        "body": [
            "#Requires AutoHotkey v2.0",
            "#SingleInstance Force",
            "SendMode 'Input'",
            "SetWorkingDir A_ScriptDir",
            "",
            ""
        ]
    }
}
```

![ahkpp-2](https://github.com/mark-wiemer-org/ahkpp/assets/1332009/b9f4992f-5edf-4d19-9335-8af9d8373460)

But, wouldn't it be possible to just add the ability to open and edit the default template/snippets files from the AHKpp settings?
`...\.vscode\extensions\mark-wiemer.vscode-autohotkey-plus-plus-5.0.3\language\snippetsV1.json`
`...\.vscode\extensions\mark-wiemer.vscode-autohotkey-plus-plus-5.0.3\language\snippetsV2.json`

**Comment by mark-wiemer** [comment]
How exactly are you opening the picker that you screenshotted? For me, the default user snippets look like this:

![image](https://github.com/mark-wiemer-org/ahkpp/assets/7833360/03f0a25e-0b30-4ead-9b26-d4549f6c8b71)

The name of the file shouldn't matter, just the name of the snippet itself. Would you mind experimenting without changing the file name and changing the name of your snippet? You should use a snippet name **different** than `AhkV2Template` for best results--by default the extension will see that it's the default snippet name and use the default snippet :/

**Comment by mark-wiemer** [comment]
The "redundant?" snippet files you mention are necessary: they load the pre-built snippets for library functions--you can see them at https://github.com/mark-wiemer-org/ahkpp/tree/main/language

**Reply by mark-wiemer** [reply]
Oops, the snippet files you screenshotted here are not the `main/language` files, they're instead files that hold user-defined snippets that will only work in those languages. They're not redundant files, just similarly named ones.

**Comment by fade2gray** [comment]
OK, I can see where I went wrong now. I selected `New Global Snippets file...` instead of the `ahk2` option, because when selecting either the `ahk` or `ahk2` options I thought they may have been redundant artefacts, because they didn't include the default templates (which I have since discovered are in `snippetsV1.json` and `snippetsV2.json`).

In my defence, though, I think the [Readme](https://github.com/mark-wiemer-org/ahkpp/blob/main/docs/TemplateSnippet.md) could be a little clearer, e.g. ...

---
You can customize the template snippet to fit your needs:

1. In VS Code, open the command palette (F1) and go to "Snippets: Configure User Snippets".
2. From the option picker, select `ahk` for a V1 template, or `ahk2` for a V2 template.
3. Create a new template snippet with a unique name, for example:

```JSON
    "CustomAhkTemplate": {
        "body": [
            "; My new script with no directives",
            "; Here is a second comment line",
            ""
        ]
    }
```

Update your VS Code settings for AHK > File > "Template snippet name" to the name of your snippet (CustomAhkTemplate)

**Reply by mark-wiemer** [reply]
Hm, I know where I went wrong: I created a `New Global Snippets file` years ago and forgot that I had taken that action. I thought everyone had a global snippets file by default! I'll update the README, thanks for this :)

**Reply by mark-wiemer** [reply]
Ref #410
